ananse/enhancer_binding.py

```diff
@@ -0,0 +1,497 @@
+
+from ananse.utils import (
+    bed_sort,
+    bed_merge,
+    count_reads,
+    bam_index,
+    bam_merge,
+    bam_sort,
+)
+from ananse.distributions import Distributions
+
+
+class CombineBedFiles:
+    def __init__(self, genome, peakfiles, verbose=True):
+        self.genome = genome
+        self.list_of_peakfiles = (
+            peakfiles if isinstance(peakfiles, list) else [peakfiles]
+        )
+        self.verbose = verbose
+
+    @staticmethod
+    def is_narrowpeak(bed, check_values=True):
+        """
+        Check BED type by column count.
+        Check if peak values are not all zeroes unless check_values is False.
+
+        Accepts a BED file (including narrowPeak, broadPeak, etc.)
+        Returns bool
+        """
+        with open(bed) as b:
+            for line in b:
+                if line.startswith("#"):
+                    continue
+                line = line.split("\t")
+                cols = len(line)
+                break
+
+        # narrowPeak has 10 columns
+        # and the peak column is >= 0
+        if cols != 10 or int(line[9]) < 0:
+            return False
+
+        if not check_values:
+            return True
+
+        # check if the peak values aren't all zeroes
+        summit_values = 0
+        sample_size = 20  # check an arbitrary number of lines
+        with open(bed) as b:
+            for n, line in enumerate(b):
+                if line.startswith("#"):
+                    continue
+                line = line.split("\t")
+                peak_val = int(line[9])
+
+                # value must be >=0
+                if peak_val < 0:
+                    return False
+                summit_values += peak_val
+                if n >= sample_size:
+                    break
+
+        if summit_values > 0:
+            return True
+        return False
+
+    @staticmethod
+    def bed_resize(
+        genome,
+        bed_in,
+        bed_out,
+        width=200,
+        narrowpeak=False,
+        fix_outliers=False,
+        output_bed3=True,
+        verbose=True,
+    ):
+        """
+        Set bed region width.
+
+        If the input bed is a narrowPeak file (narrowpeak=True),
+        center region on the summit (start+peak).
+        Otherwise center on the middle of the region.
+
+        If fix_outliers is set to True, shift regions to fit their chromosomes.
+        Otherwise drop these regions.
+
+        If output_bed3 is set to False, output the whole bed file.
+        """
+        half_seqlen = width // 2
+        chrom_sizes = genomepy.Genome(genome).sizes
+        missing_chrm = []
+
+        if narrowpeak:
+
+            def get_summit(_start, _, summit_offset):
+                return _start + int(summit_offset)
+
+            summit_col = 9
+        else:
+
+            def get_summit(_start, _end, _):
+                return (_start + _end) // 2
+
+            summit_col = 0  # unused
+
+        with open(bed_in) as old, open(bed_out, "w") as new:
```

Comment:
Why not pybedtools.slop?
https://daler.github.io/pybedtools/autodocs/pybedtools.bedtool.BedTool.slop.html

Reply:
slop increases/decreases the region size. This sets the region size to a fixed value (same as Quan's method).